README.md
# noaar
<!-- badges: start -->
[![R build status](https://github.com/dragosmg/noaar/workflows/R-CMD-check/badge.svg)](https://github.com/dragosmg/noaar/actions)
 [![Coverage status](https://codecov.io/gh/dragosmg/noaar/branch/master/graph/badge.svg)](https://codecov.io/github/dragosmg/noaar?branch=master)
<!-- badges: end -->

Package for working with the NOAA Significant Earthquakes dataset. For more details see the vignette.



NEWS.md
# noaar 0.1.0

• test 
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
Sys.setenv("R_TESTS" = "")
library(testthat)
library(noaar)

test_check("noaar")


tests/testthat/test_all.R
context("Test functions in the package")

filename <- system.file("extdata/signif.txt.tsv", package = "noaar")
data <- readr::read_delim(filename, delim = "\t")

test_that("eq_clean_data returns data frame", {
    expect_is(eq_clean_data(data), "data.frame")
})

test_that("eq_clean_data$DATE is Date type", {
    expect_is(eq_clean_data(data)$DATE, "Date")
})

test_that("eq_clean_data returns numeric coordinates", {
    expect_is(eq_clean_data(data)$LATITUDE, "numeric")
    expect_is(eq_clean_data(data)$LONGITUDE, "numeric")
})

test_that("eq_location_clean returns a data frame", {
    expect_is(eq_location_clean(data), "data.frame")
})

test_that("geom_timeline returns ggplot object", {
    g <- data %>% eq_clean_data() %>%
        dplyr::filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
        ggplot2::ggplot(ggplot2::aes(x = DATE,
                                     y = COUNTRY,
                                     color = as.numeric(TOTAL_DEATHS),
                                     size = as.numeric(EQ_PRIMARY)
        )) +
        geom_timeline()
    expect_is(g, "ggplot")
})

test_that("geom_timeline_label returns ggplot object", {
    g <- data %>% eq_clean_data() %>%
        dplyr::filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
        ggplot2::ggplot(ggplot2::aes(x = DATE,
                                     y = COUNTRY,
                                     color = as.numeric(TOTAL_DEATHS),
                                     size = as.numeric(EQ_PRIMARY)
        )) +
        geom_timeline_label(aes(label = LOCATION_NAME))
    expect_is(g, "ggplot")
})

test_that("theme_timeline returns ggplot object", {
    g <- data %>% eq_clean_data() %>%
        dplyr::filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
        ggplot2::ggplot(ggplot2::aes(x = DATE,
                                     y = COUNTRY,
                                     color = as.numeric(TOTAL_DEATHS),
                                     size = as.numeric(EQ_PRIMARY)
        )) +
        theme_timeline()
    expect_is(g, "ggplot")
})

test_that("eq_map returns leaflet object", {
    l <- data %>%
        eq_clean_data() %>%
        dplyr::filter(COUNTRY == "MEXICO" & lubridate::year(DATE) >= 2000) %>%
        dplyr::mutate(popup_text = eq_create_label(.)) %>%
        eq_map(annot_col = "popup_text")
    expect_is(l, "leaflet")
})

test_that("eq_create_label returns character vector", {
    expect_is(eq_create_label(data), "character")
})

test_that("eq_create_label has the same length as input df", {
    expect_equal(length(eq_create_label(data)), dim(data)[1])
})


tests/testthat/test-eq-clean-data.R
context("eq_clean_data")

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})


R/maps.R

#' Leaflet map of earthquakes
#'
#' This function creates a \code{leaflet} map of selected earthquakes based on
#' input NOAA earthquake cleaned data.
#'
#' @param data A data frame containing cleaned NOAA earthquake data
#' @param annot_col A character. The name of the column in the data that should
#' be used as descriptor.
#'
#' @return A leaflet map with earthquakes and annotations.
#' @export
#'
#' @importFrom leaflet leaflet addTiles addCircleMarkers
#'
#' @examples
#' \dontrun{
#' eq_map(data, annot_col = "LOCATION_NAME")
#' }
eq_map <- function(data, annot_col) {

    m <- leaflet::leaflet() %>%
        leaflet::addTiles() %>%
        leaflet::addCircleMarkers(lng = data$LONGITUDE, lat = data$LATITUDE,
                                  radius = data$EQ_PRIMARY, weight = 1,
                                  popup = data[[annot_col]])

    m
}


#' Creates a label for leaflet map
#'
#' This function creates a label for the \code{leaflet} map based on location
#' name, magnitude and casualties from NOAA earthquake data
#'
#' @param data A data frame containing cleaned NOAA earthquake data
#'
#' @return A character vector with labels
#'
#' @details The input \code{data.frame} needs to include columns LOCATION_NAME,
#' EQ_PRIMARY and TOTAL_DEATHS with the earthquake location, magintude and
#' total casualties respectively.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' eq_create_label(data)
#' }
eq_create_label <- function(data) {
    popup_text <- with(data, {
        part1 <- ifelse(is.na(LOCATION_NAME), "",
                        paste("<strong>Location:</strong>",
                              LOCATION_NAME))
        part2 <- ifelse(is.na(EQ_PRIMARY), "",
                        paste("<br><strong>Magnitude</strong>",
                              EQ_PRIMARY))
        part3 <- ifelse(is.na(TOTAL_DEATHS), "",
                        paste("<br><strong>Total deaths:</strong>",
                              TOTAL_DEATHS))
        paste0(part1, part2, part3)
    })
}



R/leaflet.R

#' Leaflet map of earthquakes
#'
#' This function creates a \code{leaflet} map of selected earthquakes based on
#' input NOAA earthquake cleaned data.
#'
#' @param data A data frame containing cleaned NOAA earthquake data
#' @param annot_col A character. The name of the column in the data that should
#' be used as descriptor.
#'
#' @return A leaflet map with earthquakes and annotations.
#' @export
#'
#' @importFrom leaflet leaflet addTiles addCircleMarkers
#'
#' @examples
#' \dontrun{
#' eq_map(data, annot_col = "LOCATION_NAME")
#' }
eq_map <- function(data, annot_col) {

    m <- leaflet::leaflet() %>%
        leaflet::addTiles() %>%
        leaflet::addCircleMarkers(lng = data$LONGITUDE, lat = data$LATITUDE,
                                  radius = data$EQ_PRIMARY, weight = 1,
                                  popup = data[[annot_col]])

    m
}


#' Creates a label for leaflet map
#'
#' This function creates a label for the \code{leaflet} map based on location
#' name, magnitude and casualties from NOAA earthquake data
#'
#' @param data A data frame containing cleaned NOAA earthquake data
#'
#' @return A character vector with labels
#'
#' @details The input \code{data.frame} needs to include columns LOCATION_NAME,
#' EQ_PRIMARY and TOTAL_DEATHS with the earthquake location, magintude and
#' total casualties respectively.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' eq_create_label(data)
#' }
eq_create_label <- function(data) {
    popup_text <- with(data, {
        part1 <- ifelse(is.na(LOCATION_NAME), "",
                        paste("<strong>Location:</strong>",
                              LOCATION_NAME))
        part2 <- ifelse(is.na(EQ_PRIMARY), "",
                        paste("<br><strong>Magnitude</strong>",
                              EQ_PRIMARY))
        part3 <- ifelse(is.na(TOTAL_DEATHS), "",
                        paste("<br><strong>Total deaths:</strong>",
                              TOTAL_DEATHS))
        paste0(part1, part2, part3)
    })
}


R/data_clean.R
#' Cleans source data from NOAA
#'
#' \code{eq_clean_data} cleans date, latitude and longitude, and location name
#' from the source NOAA data
#'
#' @param data A data frame with raw data obtained from NOAA website (see below)
#'
#' @return A data frame with cleaned date, latitude, longitude and location
#' columns
#'
#' @details The function requires raw date obtained from NOAA site
#' \url{https://www.ngdc.noaa.gov/nndc/struts/form?t=101650&s=1&d=1}. It adds
#' a column DATE with cleaned date (Date format), transforms LATITUDE and
#' LONGITUDE columns as numeric objects and transforms LOCATION_NAME by removing
#' the country and transforming to title case.
#'
#' @examples
#' \dontrun{
#' data <- readr::read_delim("earthquakes.tsv.gz", delim = "\t")
#' clean_data <- eq_clean_data(data)
#' }
#'
#' @importFrom dplyr %>% mutate select
#' @importFrom lubridate ymd
#' @importFrom stringr str_pad
#'
#' @export
eq_clean_data <- function(data) {
    data <- data %>%
        dplyr::mutate_(
            year_fix = ~stringr::str_pad(as.character(abs(YEAR)), width = 4,
                                         side = "left", pad = "0"),
            date_paste = ~paste(year_fix, MONTH, DAY, sep = "-"),
            DATE = ~lubridate::ymd(date_paste, truncated = 2)) %>%
        dplyr::select_(quote(-year_fix), quote(-date_paste))

    lubridate::year(data$DATE) <- data$YEAR

    data <- data %>%
        dplyr::mutate_(LATITUDE = ~as.numeric(LATITUDE),
                       LONGITUDE = ~as.numeric(LONGITUDE))

    data <- eq_location_clean(data)

    data
}


#' Cleans earthquake location data
#'
#' @param data A data frame with raw data obtained from NOAA website
#'
#' @return A data frame with cleaned LOCATION_NAME column
#'
#' @details This function transforms NOAA data frame LOCATION_NAME column by
#' trimming the country name (if applicable) and converting to title case
#'
#' @note The function is not exported
#'
#' @examples
#' \dontrun{
#' data <- readr::read_delim("earthquakes.tsv.gz", delim = "\t")
#' clean_data <- eq_location_clean(data)
#' }
#'
#' @importFrom dplyr %>% mutate
#' @importFrom stringr str_replace str_trim str_to_title
eq_location_clean <- function(data) {
    data <- data %>%
        dplyr::mutate_(LOCATION_NAME = ~LOCATION_NAME %>%
                           stringr::str_replace(paste0(COUNTRY, ":"), "") %>%
                           stringr::str_trim("both") %>%
                           stringr::str_to_title())
    data
}


R/eq_clean_data.R
#' Cleans source data from NOAA
#'
#' \code{eq_clean_data} cleans date, latitude and longitude, and location name
#' from the source NOAA data
#'
#' @param data A data frame with raw data obtained from NOAA website (see below)
#'
#' @return A data frame with cleaned date, latitude, longitude and location
#' columns
#'
#' @details The function requires raw date obtained from NOAA site
#' \url{https://www.ngdc.noaa.gov/nndc/struts/form?t=101650&s=1&d=1}. It adds
#' a column DATE with cleaned date (Date format), transforms LATITUDE and
#' LONGITUDE columns as numeric objects and transforms LOCATION_NAME by removing
#' the country and transforming to title case.
#'
#' @examples
#' \dontrun{
#' data <- readr::read_delim("earthquakes.tsv.gz", delim = "\t")
#' clean_data <- eq_clean_data(data)
#' }
#'
#' @importFrom dplyr %>% mutate select
#' @importFrom lubridate ymd
#' @importFrom stringr str_pad
#'
#' @export
eq_clean_data <- function(data) {
    data <- data %>%
        dplyr::mutate_(
            year_fix = ~stringr::str_pad(as.character(abs(YEAR)), width = 4,
                                         side = "left", pad = "0"),
            date_paste = ~paste(year_fix, MONTH, DAY, sep = "-"),
            DATE = ~lubridate::ymd(date_paste, truncated = 2)) %>%
        dplyr::select_(quote(-year_fix), quote(-date_paste))

    lubridate::year(data$DATE) <- data$YEAR

    data <- data %>%
        dplyr::mutate_(LATITUDE = ~as.numeric(LATITUDE),
                       LONGITUDE = ~as.numeric(LONGITUDE))

    data <- eq_location_clean(data)

    data
}


R/geom_timeline.R
#' Earthquakes timeline
#'
#' This geom plots a timeline of earthquakes between 2 dates, with options to
#' map different variables to other aesthetics. The x aesthetic is a date and
#' the y aesthetic is optional and needs to be a factor indicating some
#' stratification (e.g. country, magnitude, etc).
#'
#' @inheritParams ggplot2::geom_point
#'
#' @return a ggplot2 graphical object
#' @export
#'
#' @importFrom ggplot2 layer
#'
#' @examples
#' \dontrun{
#'#' data %>% eq_clean_data() %>%
#'    filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
#'    ggplot(aes(x = DATE,
#'               y = COUNTRY)) +
#'    geom_timeline()
#' }
geom_timeline <- function(mapping = NULL, data = NULL, stat = "identity",
                          position = "identity", ..., na.rm = FALSE,
                          show.legend = NA, inherit.aes = TRUE) {
    ggplot2::layer(
        geom = GeomTimeline, mapping = mapping, data = data, stat = stat,
        position = position, show.legend = show.legend,
        inherit.aes = inherit.aes, params = list(na.rm = na.rm, ...)
    )
}

#' @importFrom ggplot2 aes draw_key_point ggproto
#' @importFrom grid pointsGrob linesGrob gList gpar
#' @importFrom scales alpha
GeomTimeline <-
    ggplot2::ggproto(
        "GeomTimeline", ggplot2::Geom,
        required_aes = c("x"),
        default_aes = ggplot2::aes(colour = "grey", size = 1.5, alpha = 0.5,
                                   shape = 21, fill = "grey", stroke = 0.5),
        draw_key = ggplot2::draw_key_point,
        draw_panel = function(data, panel_scales, coord) {

            if (!("y" %in% colnames(data))) {
                data$y <- 0.15
            }

            coords <- coord$transform(data, panel_scales)

            points <- grid::pointsGrob(
                coords$x, coords$y,
                pch = coords$shape, size = unit(coords$size / 4, "char"),
                gp = grid::gpar(
                    col = scales::alpha(coords$colour, coords$alpha),
                    fill = scales::alpha(coords$colour, coords$alpha)
                )
            )
            y_lines <- unique(coords$y)

            lines <- grid::polylineGrob(
                x = unit(rep(c(0, 1), each = length(y_lines)), "npc"),
                y = unit(c(y_lines, y_lines), "npc"),
                id = rep(seq_along(y_lines), 2),
                gp = grid::gpar(col = "grey",
                                lwd = .pt)
            )

            grid::gList(points, lines)
        }
    )


R/eq_location_clean.R
#' Cleans earthquake location data
#'
#' @param data A data frame with raw data obtained from NOAA website
#'
#' @return A data frame with cleaned LOCATION_NAME column
#'
#' @details This function transforms NOAA data frame LOCATION_NAME column by
#' trimming the country name (if applicable) and converting to title case
#'
#' @note The function is not exported
#'
#' @examples
#' \dontrun{
#' data <- readr::read_delim("earthquakes.tsv.gz", delim = "\t")
#' clean_data <- eq_location_clean(data)
#' }
#'
#' @importFrom dplyr %>% mutate
#' @importFrom stringr str_replace str_trim str_to_title
eq_location_clean <- function(data) {
    data <- data %>%
        dplyr::mutate_(LOCATION_NAME = ~LOCATION_NAME %>%
                           stringr::str_replace(paste0(COUNTRY, ":"), "") %>%
                           stringr::str_trim("both") %>%
                           stringr::str_to_title())
    data
}


R/geom_timeline_label.R
#' Timeline labels of earthquakes
#'
#' @description This geom plots timeline labels of earthquakes. It assumes that
#' \code{geom_timeline} was used to create the timelines
#'
#' @inheritParams ggplot2::geom_text

#' @param n_max An integer. If used, it only plots the labels for the
#' \code{n_max} largest earthquakes in the selected group in the timeline
#'
#' @details The function plots timeline labels of earthquakes based on cleaned
#' NOAA data. It should be used with combination with \code{geom_timeline}. The
#' required aesthetics for this geom is \code{label} that should contain
#' string for labeling each data point.
#'
#' @export
#'
#' @importFrom ggplot2 layer
#'
#' @examples
#' \dontrun{
#' data %>% eq_clean_data() %>%
#'    filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
#'    ggplot(aes(x = DATE,
#'               y = COUNTRY,
#'               color = as.numeric(TOTAL_DEATHS),
#'               size = as.numeric(EQ_PRIMARY)
#'    )) +
#'    geom_timeline() +
#'    geom_timeline_label(aes(label = LOCATION_NAME), n_max = 5) +
#'    theme_timeline() +
#'    labs(size = "Richter scale value", color = "# deaths")
#' }
geom_timeline_label <- function(mapping = NULL, data = NULL, stat = "identity",
                                position = "identity", ..., na.rm = FALSE,
                                n_max = NULL, show.legend = NA,
                                inherit.aes = TRUE) {

    ggplot2::layer(
        geom = GeomTimelineLabel, mapping = mapping,
        data = data, stat = stat, position = position,
        show.legend = show.legend, inherit.aes = inherit.aes,
        params = list(na.rm = na.rm, n_max = n_max, ...)
    )
}

#' @importFrom ggplot2 draw_key_blank
#' @importFrom dplyr %>% group_by top_n ungroup
#' @importFrom grid gpar linesGrob textGrob gList
GeomTimelineLabel <-
    ggplot2::ggproto(
        "GeomTimelineLabel", ggplot2::Geom,
        required_aes = c("x", "label"),
        draw_key = ggplot2::draw_key_blank,
        setup_data = function(data, params) {
            if (!is.null(params$n_max)) {
                if (!("size" %in% colnames(data))) {
                    stop(paste("'size' aesthetics needs to be",
                               "provided when 'n_max' is defined."))
                }
                data <- data %>%
                    dplyr::group_by_("group") %>%
                    dplyr::top_n(params$n_max, size) %>%
                    dplyr::ungroup()
            }
            data
        },
        draw_panel = function(data, panel_scales, coord, n_max) {

            if (!("y" %in% colnames(data))) {
                data$y <- 0.15
            }

            coords <- coord$transform(data, panel_scales)
            n_grp <- length(unique(data$group))
            offset <- 0.2 / n_grp

            lines <- grid::polylineGrob(
                x = unit(c(coords$x, coords$x), "npc"),
                y = unit(c(coords$y, coords$y + offset), "npc"),
                id = rep(1:dim(coords)[1], 2),
                gp = grid::gpar(
                    col = "grey"
                )
            )

            names <- grid::textGrob(
                label = coords$label,
                x = unit(coords$x, "npc"),
                y = unit(coords$y + offset, "npc"),
                just = c("left", "bottom"),
                rot = 45
            )

            grid::gList(lines, names)
        }
    )


R/geoms.R

#' Timeline of earthquakes
#'
#' @description This geom plots a timeline of earthquakes in one line with
#' options to group by country, color by number of casualties and size by scale
#'
#' @inheritParams ggplot2::geom_point
#'
#' @details The function plots a timeline of earthquakes based on cleaned NOAA
#' data. It requires \code{x} aesthetics. An optional \code{y} aesthetics can
#' be used to group data by a selected variable (for example country).
#' @export
#'
#' @importFrom ggplot2 layer
#'
#' @examples
#' \dontrun{
#' data %>% eq_clean_data() %>%
#'    filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
#'    ggplot(aes(x = DATE,
#'               y = COUNTRY,
#'               color = as.numeric(TOTAL_DEATHS),
#'               size = as.numeric(EQ_PRIMARY)
#'    )) +
#'    geom_timeline() +
#'    theme_timeline() +
#'    labs(size = "Richter scale value", color = "# deaths")
#' }
geom_timeline <- function(mapping = NULL, data = NULL, stat = "identity",
                          position = "identity", na.rm = FALSE,
                          show.legend = NA, inherit.aes = TRUE, ...) {

    ggplot2::layer(
        geom = GeomTimeline, mapping = mapping,
        data = data, stat = stat, position = position,
        show.legend = show.legend, inherit.aes = inherit.aes,
        params = list(na.rm = na.rm, ...)
    )
}

#' @importFrom ggplot2 aes draw_key_point
#' @importFrom grid pointsGrob linesGrob gList gpar
#' @importFrom scales alpha
GeomTimeline <-
    ggplot2::ggproto(
        "GeomTimeline", ggplot2::Geom,
        required_aes = c("x"),
        default_aes = ggplot2::aes(colour = "grey", size = 1.5, alpha = 0.5,
                                   shape = 21, fill = "grey", stroke = 0.5),
        draw_key = ggplot2::draw_key_point,
        draw_panel = function(data, panel_scales, coord) {

            if (!("y" %in% colnames(data))) {
                data$y <- 0.15
            }

            coords <- coord$transform(data, panel_scales)

            points <- grid::pointsGrob(
                coords$x, coords$y,
                pch = coords$shape, size = unit(coords$size / 4, "char"),
                gp = grid::gpar(
                    col = scales::alpha(coords$colour, coords$alpha),
                    fill = scales::alpha(coords$colour, coords$alpha)
                )
            )
            y_lines <- unique(coords$y)

            lines <- grid::polylineGrob(
                x = unit(rep(c(0, 1), each = length(y_lines)), "npc"),
                y = unit(c(y_lines, y_lines), "npc"),
                id = rep(seq_along(y_lines), 2),
                gp = grid::gpar(col = "grey",
                                lwd = .pt)
            )

            grid::gList(points, lines)
        }
    )

#' Timeline labels of earthquakes
#'
#' @description This geom plots timeline labels of earthquakes. It assumes that
#' \code{geom_timeline} was used to create the timelines
#'
#' @inheritParams ggplot2::geom_text

#' @param n_max An integer. If used, it only plots the labels for the
#' \code{n_max} largest earthquakes in the selected group in the timeline
#'
#' @details The function plots timeline labels of earthquakes based on cleaned
#' NOAA data. It should be used with combination with \code{geom_timeline}. The
#' required aesthetics for this geom is \code{label} that should contain
#' string for labeling each data point.
#'
#' @export
#'
#' @importFrom ggplot2 layer
#'
#' @examples
#' \dontrun{
#' data %>% eq_clean_data() %>%
#'    filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
#'    ggplot(aes(x = DATE,
#'               y = COUNTRY,
#'               color = as.numeric(TOTAL_DEATHS),
#'               size = as.numeric(EQ_PRIMARY)
#'    )) +
#'    geom_timeline() +
#'    geom_timeline_label(aes(label = LOCATION_NAME), n_max = 5) +
#'    theme_timeline() +
#'    labs(size = "Richter scale value", color = "# deaths")
#' }
geom_timeline_label <- function(mapping = NULL, data = NULL, stat = "identity",
                                position = "identity", ..., na.rm = FALSE,
                                n_max = NULL, show.legend = NA,
                                inherit.aes = TRUE) {

    ggplot2::layer(
        geom = GeomTimelineLabel, mapping = mapping,
        data = data, stat = stat, position = position,
        show.legend = show.legend, inherit.aes = inherit.aes,
        params = list(na.rm = na.rm, n_max = n_max, ...)
    )
}

#' @importFrom ggplot2 draw_key_blank
#' @importFrom dplyr %>% group_by top_n ungroup
#' @importFrom grid gpar linesGrob textGrob gList
GeomTimelineLabel <-
    ggplot2::ggproto(
        "GeomTimelineLabel", ggplot2::Geom,
        required_aes = c("x", "label"),
        draw_key = ggplot2::draw_key_blank,
        setup_data = function(data, params) {
            if (!is.null(params$n_max)) {
                if (!("size" %in% colnames(data))) {
                    stop(paste("'size' aesthetics needs to be",
                               "provided when 'n_max' is defined."))
                }
                data <- data %>%
                    dplyr::group_by_("group") %>%
                    dplyr::top_n(params$n_max, size) %>%
                    dplyr::ungroup()
            }
            data
        },
        draw_panel = function(data, panel_scales, coord, n_max) {

            if (!("y" %in% colnames(data))) {
                data$y <- 0.15
            }

            coords <- coord$transform(data, panel_scales)
            n_grp <- length(unique(data$group))
            offset <- 0.2 / n_grp

            lines <- grid::polylineGrob(
                x = unit(c(coords$x, coords$x), "npc"),
                y = unit(c(coords$y, coords$y + offset), "npc"),
                id = rep(1:dim(coords)[1], 2),
                gp = grid::gpar(
                    col = "grey"
                )
            )

            names <- grid::textGrob(
                label = coords$label,
                x = unit(coords$x, "npc"),
                y = unit(coords$y + offset, "npc"),
                just = c("left", "bottom"),
                rot = 45
            )

            grid::gList(lines, names)
        }
    )

#' Theme for better timeline visualization in ggplot2
#'
#' @description  This is a simple theme that makes \code{\link{geom_timeline}}
#' look better.
#'
#' @examples
#' \dontrun{
#' data %>% eq_clean_data() %>%
#' filter(COUNTRY %in% c("GREECE", "ITALY"), YEAR > 2000) %>%
#'    ggplot(aes(x = DATE,
#'               y = COUNTRY,
#'               color = as.numeric(TOTAL_DEATHS),
#'               size = as.numeric(EQ_PRIMARY)
#'    )) +
#'    geom_timeline() +
#'    theme_timeline()
#' }
#'
#' @importFrom ggplot2 theme element_blank element_line
#'
#' @export
#'
theme_timeline <- function() {
    ggplot2::theme(
        plot.background = ggplot2::element_blank(),
        panel.background = ggplot2::element_blank(),
        legend.key = ggplot2::element_blank(),
        axis.title.y = ggplot2::element_blank(),
        axis.line.x = ggplot2::element_line(size = 1),
        axis.ticks.y = ggplot2::element_blank(),
        legend.position = "bottom"
    )
}
